ssl/s3_lib.c

```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha is at the top of the client preference list,
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, reprioritize... */
+                new_chacha = sk_SSL_CIPHER_new_null();
```

Comment:
There is now sk_\*_reserve that should be preferred to use here. [And it would make sense to check for its return value.] Though if goal is to back-port it, then of course one can't use it. But one can and probably should add it afterward. Or as separate commit, so that one could be cherry-picked, but not the other...

Reply:
Done as a separate commit